libraries/AP_RCProtocol/AP_RCProtocol_FPort2.cpp

```diff
@@ -0,0 +1,264 @@
+    buf[2] = telem_data.frame;
+    buf[3] = telem_data.appid & 0xFF;
+    buf[4] = telem_data.appid >> 8;
+    memcpy(&buf[5], &telem_data.data, 4);
+    buf[9] = crc_sum8(&buf[1], 8);
+    
+    uart->write(buf, sizeof(buf));
+
+    // get fresh telem_data in the next call
+    telem_data.available = false;
+#endif
+}
+
+/*
+  process a FPort2 input pulse of the given width
+ */
+void AP_RCProtocol_FPort2::process_pulse(uint32_t width_s0, uint32_t width_s1)
+{
+    if (have_UART()) {
+        // if we can use a UART we would much prefer to, as it allows
+        // us to send SPORT data out
+        return;
+    }
+    uint32_t w0 = width_s0;
+    uint32_t w1 = width_s1;
+    if (inverted) {
+        w0 = saved_width;
+        w1 = width_s0;
+        saved_width = width_s1;
+    }
+    uint8_t b;
+    if (ss.process_pulse(w0, w1, b)) {
+        _process_byte(ss.get_byte_timestamp_us(), b);
+    }
+}
+
+// support byte input
+void AP_RCProtocol_FPort2::_process_byte(uint32_t timestamp_us, uint8_t b)
+{
+    const bool have_frame_gap = (timestamp_us - byte_input.last_byte_us >= 2000U);
+
+    byte_input.last_byte_us = timestamp_us;
+
+    if (have_frame_gap) {
+        // if we have a frame gap then this must be the start of a new
+        // frame
+        byte_input.ofs = 0;
+    }
+
+    if (byte_input.ofs == 0) {
+        switch (b) {
+        case FRAME_LEN_8:
+            byte_input.control_len = 16;
+            chan_count = 8;
+            byte_input.is_downlink = false;
+            break;
+        case FRAME_LEN_16:
+            byte_input.control_len = 27;
+            chan_count = 16;
+            byte_input.is_downlink = false;
+            break;
+        case FRAME_LEN_24:
+            byte_input.control_len = 37;
+            chan_count = 24;
+            byte_input.is_downlink = false;
+            break;
+        case FRAME_LEN_DOWNLINK:
+            byte_input.control_len = 10;
+            byte_input.is_downlink = true;
+            break;
+        default:
+            // definately not FPort2, missing header byte
+            return;
+        }
+    }
+
+    if (byte_input.ofs == 1) {
+        if (!byte_input.is_downlink && b != FRAME_TYPE_CHANNEL) {
+            // not channel data
+            byte_input.ofs = 0;
+            return;
+        }
+    }
+
+    byte_input.buf[byte_input.ofs++] = b;
+
+    const FPort2_Frame *frame = (const FPort2_Frame *)&byte_input.buf[0];
+
+    if (byte_input.control_len > 2 && byte_input.ofs == byte_input.control_len) {
+        if (!byte_input.is_downlink) {
+            log_data(AP_RCProtocol::FPORT2, timestamp_us, byte_input.buf, byte_input.ofs);
+            if (check_checksum()) {
+                decode_control(*frame);
+            }
+        } else {
+            // downlink packet
+            if (check_checksum()) {
+                decode_downlink(*frame);
+            }
+        }
+        goto reset;
+    }
+    if (byte_input.ofs == sizeof(byte_input.buf)) {
```

Comment:
```suggestion
    if (byte_input.ofs >= sizeof(byte_input.buf)) {
```

just for defensiveness

Reply:
done